adr/ADR-20.md

```diff
@@ -0,0 +1,73 @@
+This mechanism needs to be supported for materialized JS views as well.
+
+## Overview 
+
+Each JetStream only listens to default API subjects with the prefix `$JS.API`.
+Thus, when the client uses `$JS.API`, it communicates with the JetStream, local to it's account and domain.
+
+To avoid traffic going to some other JetStream the following mechanisms are in place:
+1. Account: Since the API has to be imported with an altered prefix, the request will not cross account boundaries. 
+    In a JetStream enabled account, an import without a prefix set will result in an error as JetStream is imported as well and we error on import overlaps. 
+2. Domain: On leaf node connections, the nats server adds in denies for `$JS.API.>`.
+
+When the client library sets an API prefix, all API subjects, the client publishes and subscribes to start with that instead of `$JS.API`.
+As a result the API traffic will not end up in the local JetStream as that alway only ever subscribes to `$JS.API....`
+
+As messages and subscriptions cross boundaries the following happens:
+1. Accounts: When crossing the import, the API prefix is stripped and replaced with `$JS.API`.
+2. Domain: When crossing the leaf node connection between domains, an automatically inserted mapping strips the API prefix and replaces it with `$JS.API`.
+
+This JetStream disambiguation mechanism needs to be added to materialized views such as KV and object store as well.
+Specifically, we need to tag along on the same API prefix. 
+Setting different values to reach the same JetStream is a non starter.
+
+## Design
+
+The first token of any API or materialized view is considered the default API prefix and will mean have local semantics.
+Thus, for the concrete views we treat the tokens `$KV` and `$OBJ` as the respective default API prefixes. 
+For publishes and subscribes the API just replaces the first token with the specified (non default) prefix. 
+
+To share API access across accounts this is sufficient as account export/import takes care of the rest. 
+To access an API in the same account but different domain, the `nats-server` maintaining a leaf node connection will add in the appropriate mappings from domain specific API to local API and deny local API traffic.
+
+## KV Example
+
+Assume the API prefix for JetStream has been set to `JS.from-acc1`.
+For JetStream specific API calls, the local API prefix `$JS.API` has to be replaced with `JS.from-acc1`.
+For the KV API we prefix `$KV` with `JS.from-acc1`, resulting in `JS.from-acc1.$KV`. 
+Thus, in order to put `key` in the bin `bin`, we send to `JS.from-acc1.$KV.bin.key` 
+
+When crossing the account boundaries, this is then translated back to `$KV.bin.key`.
+Thus, the underlying stream still needs to be created with a subscription to `$KV.bin.key`. 
+
+Domains are just a special case of API prefixes and will work the same way.
+The API prefix `$JS.<domain-name>.API` will lead to `$JS.<domain-name>.API.$KV.bin.key`.
```

Comment:
But for domains the client libraries would then need to parse the new domain subject as the rewriting wont happen right? Or will you also rewrite there as per account boundary?

Reply:
Rewriting happens for domains as well. same as line 33. But I'll add a sentence for clarification.